src/pep257.py

```diff
@@ -854,68 +1233,40 @@ def filter(self, record):
     log.addHandler(stderr_handler)
 
 
-def get_checked_error_codes(options):
-    codes = set(ErrorRegistry.get_error_codes())
-    if options.ignore is not None:
-        checked_codes = codes - set(options.ignore.split(','))
-    elif options.select is not None:
-        checked_codes = set(options.select.split(','))
-    elif options.convention is not None:
-        checked_codes = getattr(conventions, options.convention)
-    else:
-        checked_codes = conventions.pep257
-    checked_codes -= set(options.add_ignore.split(','))
-    checked_codes |= set(options.add_select.split(','))
-    return checked_codes - set('')
-
-
-def validate_options(options):
-    mutually_exclusive = ('ignore', 'select', 'convention')
-    for opt1, opt2 in itertools.permutations(mutually_exclusive, 2):
-        if getattr(options, opt1) and getattr(options, opt2):
-            log.error('Cannot pass both {0} and {1}. They are '
-                      'mutually exclusive.'.format(opt1, opt2))
-            return False
-    if options.convention and options.convention not in conventions:
-        log.error("Illegal convention '{0}'. Possible conventions: {1}"
-                  .format(options.convention, ', '.join(conventions.keys())))
-        return False
-    return True
-
-
 def run_pep257():
     log.setLevel(logging.DEBUG)
-    opt_parser = get_option_parser()
-    # setup the logger before parsing the config file, so that command line
-    # arguments for debug / verbose will be printed.
-    options, arguments = opt_parser.parse_args()
-    setup_stream_handlers(options)
-    # We parse the files before opening the config file, since it changes where
-    # we look for the file.
-    options = get_options(arguments, opt_parser)
-    if not validate_options(options):
+    conf = ConfigurationParser()
+    setup_stream_handlers(conf.get_default_run_configuration())
+
+    try:
+        conf.parse()
+    except IllegalConfiguration:
         return INVALID_OPTIONS_RETURN_CODE
-    # Setup the handler again with values from the config file.
-    setup_stream_handlers(options)
 
-    collected = collect(arguments or ['.'],
-                        match=re(options.match + '$').match,
-                        match_dir=re(options.match_dir + '$').match)
+    run_conf = conf.get_user_run_configuration()
+
+    # Reset the logger according to the command line arguments
+    setup_stream_handlers(run_conf)
 
     log.debug("starting pep257 in debug mode.")
 
-    Error.explain = options.explain
-    Error.source = options.source
-    collected = list(collected)
-    checked_codes = get_checked_error_codes(options)
-    errors = check(collected, select=checked_codes)
+    Error.explain = run_conf.explain
+    Error.source = run_conf.source
+
+    errors = []
+    try:
+        for filename, checked_codes in conf.get_files_to_check():
+            errors.extend(check((filename,), select=checked_codes))
+    except IllegalConfiguration:
```

Comment:
Not all configuration files are read in `config.parse()`, they are read on the go.
It's possible that during generation a new configuration file will be found and be illegal.
e.g:
A
+-- tox.ini - `select=D100`
+-- a.py
+-- B
     +-- tox.ini - `select=D100' and `ignore=D100`
     +-- b.py

Calling `pep257.py A` will raise only on the second iteration of that generator.
I'll add a comment.